Added TestSourceProcessor helper

Dr15Jones opened the pull request:
#### PR description:

Added TestSourceProcessor which is similar to TestProcessor but is specialized for testing Sources.

#### PR validation:

Code compiles and new unit tests pass.

Dr15Jones committed:
Fix incorrect name for exception

Dr15Jones committed:
Added TestSourceProcessor testing helper

This is similar to TestProcessor but specialized for testing
Sources.

Dr15Jones commented:
please test

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"078d2cb009777f98048c224d3c274f6731943abd":{"files":["FWCore/TestProcessor/interface/EventFromSource.h","FWCore/TestProcessor/interface/LuminosityBlockFromSource.h","FWCore/TestProcessor/interface/RunFromSource.h","FWCore/TestProcessor/interface/TestSourceProcessor.h","FWCore/TestProcessor/python/TestSourceProcess.py","FWCore/TestProcessor/src/TestSourceProcessor.cc","FWCore/TestProcessor/test/BuildFile.xml","FWCore/TestProcessor/test/testsourceprocessor_t.cppunit.cc"],"squashed":false,"time":1720806470},"864c800bec39c8ad54b99f42949a4859320945d8":{"files":["FWCore/TestProcessor/python/TestProcess.py"],"squashed":false,"time":1720806374}},"emoji":{},"last_seen_sha":"078d2cb009777f98048c224d3c274f6731943abd","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45443/40893

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones for master.

It involves the following packages:

- FWCore/TestProcessor (**core**)


@Dr15Jones, @makortel, @smuzaffar can you please review it and eventually sign? Thanks.
@makortel, @missirol, @wddgit this is something you requested to watch as well.
@antoniovilela, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f7a52e/40366/summary.html
**COMMIT**: 078d2cb009777f98048c224d3c274f6731943abd
**CMSSW**: CMSSW_14_1_X_2024-07-11-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/45443/40366/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test testFWCoreTestProcessor had ERRORS
</pre>

## Comparison Summary



There are some workflows for which there are errors in the baseline:
 29634.911 step 2
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially [removed 13 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-11-2300+f7a52e/63560/validateJR/logRootQA-events.log) from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-11-2300+f7a52e/63560/validateJR.html)
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3246798
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-11-2300+f7a52e/63560/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3246778
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 199 log files, 163 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-11-2300+f7a52e/63560/triggerResults): no differences found

makortel commented:
Unit test failure looks related

Dr15Jones committed:
Share one time initialization between testing frameworks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45443/40923

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #45434

cmsbuild commented:
Pull request #45443 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f7a52e/40411/summary.html
**COMMIT**: ff4d600d11051573dbd9ad8987a8d64f1e5745b4
**CMSSW**: CMSSW_14_1_X_2024-07-15-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45443/40411/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


## Comparison Summary



**Summary**:
 - You potentially [removed 7 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-15-1100+f7a52e/63595/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-15-1100+f7a52e/63595/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3345094
 - DQMHistoTests: [Total failures: 3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-15-1100+f7a52e/63595/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3345071
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-15-1100+f7a52e/63595/triggerResults): no differences found

makortel commented on FWCore/TestProcessor/src/oneTimeInitialization.h:
```diff
@@ -0,0 +1,9 @@
+#if !defined(FWCore_TestProcessor_oneTimeInitialization_h)
```
Could this be 
```suggestion
#ifndef FWCore_TestProcessor_oneTimeInitialization_h
```
?

Dr15Jones commented:
hold

additional testing using PoolSource has shown problems.

cmsbuild commented:
Pull request has been put on hold by  @Dr15Jones
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

Dr15Jones committed:
Fix meta-data and Service issues in TestSourceProcessor

Testing using PoolSource uncovered these issues.

Dr15Jones committed:
Fixed OutputModule issues in TestProcessor

Testing with PoolOutputModule showed issues with meta-data and
missing framework calls.

Dr15Jones committed:
Add unit test for Pool Output/Source

This uses the new TestSourceProcessor and changes in TestProcessor
to support OutputModules.

Dr15Jones commented:
unhold

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45443/40950

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #45434

cmsbuild commented:
Pull request #45443 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 32KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f7a52e/40458/summary.html
**COMMIT**: d08336b2957dd876306c7634dffd4bf4016147ce
**CMSSW**: CMSSW_14_1_X_2024-07-17-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45443/40458/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestGeneratorInterfacePythia8ConcurrentGeneratorFilter had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 7 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63634/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 4 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63634/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3345094
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63634/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3345074
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63634/triggerResults): no differences found

Dr15Jones commented:
The unit test failure during the test appears to be from a failed cmsDriver.py command. That could not have been caused by this PR.

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 32KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f7a52e/40461/summary.html
**COMMIT**: d08336b2957dd876306c7634dffd4bf4016147ce
**CMSSW**: CMSSW_14_1_X_2024-07-17-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/45443/40461/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


## Comparison Summary



**Summary**:
 - You potentially [removed 5 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63637/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 93 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63637/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3345094
 - DQMHistoTests: [Total failures: 2467](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63637/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3342607
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 202 log files, 165 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-07-17-1100+f7a52e/63637/triggerResults): no differences found

Dr15Jones committed:
Use ifndef to be consistent with other headers

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-45443/40967

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #45434

cmsbuild commented:
Pull request #45443 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

Dr15Jones commented:
please test

makortel commented:
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803 and 29634.911 (known non-reproducibility, but without a GH issue)

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @mandrenguyen, @antoniovilela, @sextonkennedy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

srimanob commented:
Hi @makortel 
Strange that the fix is on HGCal, but the failures are not on HGCal at all. Any possibility that somethings come between CMSSW_14_1_X_2024-07-15-2300 and now ?

makortel commented:
> Any possibility that something comes between CMSSW_14_1_X_2024-07-15-2300 and now ?

Of course that is possible, but I'd suspect more an existing issue that was hidden by the HGCal uninitialized variables. For the record, here are the PRs merged after 07-15-2300 up to 07-18-1100
* https://github.com/cms-sw/cmssw/pull/45470
* https://github.com/cms-sw/cmssw/pull/45446
* https://github.com/cms-sw/cmssw/pull/45366
* https://github.com/cms-sw/cmssw/pull/45428
* https://github.com/cms-sw/cmssw/pull/45465
* https://github.com/cms-sw/cmssw/pull/45307
* https://github.com/cms-sw/cmssw/pull/45375
* https://github.com/cms-sw/cmssw/pull/45452
* https://github.com/cms-sw/cmssw/pull/45409
* https://github.com/cms-sw/cmssw/pull/45424
* https://github.com/cms-sw/cmssw/pull/45437
* https://github.com/cms-sw/cmssw/pull/45444
* https://github.com/cms-sw/cmssw/pull/45445
* https://github.com/cms-sw/cmssw/pull/45447
* https://github.com/cms-sw/cmssw/pull/45455
* https://github.com/cms-sw/cmssw/pull/45457
* https://github.com/cms-sw/cmssw/pull/45471

rappoccio commented:
+1
